File: Java/CodesInterviews/src/main/java/org/example/google/BoatsToSavePeople.java
package org.example.google;

import java.util.Arrays;

public class BoatsToSavePeople {

    public Integer numRescueBoats(int[] people, int limit) {
        int[] sortedPeople = Arrays.stream(people).sorted().toArray();
        int smallest = 0;
        int biggest = sortedPeople.length - 1;
        int boats = 0;

        while (biggest >= smallest) {
            if (sortedPeople[smallest] + sortedPeople[biggest] <= limit) {
                biggest -= 1;
                smallest += 1;
            } else {
                biggest -= 1;
            }
            boats += 1;
        }

        return boats;
    }

    public static void main(String[] args) {
        BoatsToSavePeople boats = new BoatsToSavePeople();
        int [] people = {3, 5, 3, 4};
        int limit = 5;
        System.out.println(boats.numRescueBoats(people, limit));
    }
}

File: Kotlin/CodesInterviews/src/main/kotlin/google/BoatsToSavePeople.kt
package google

class BoatsToSavePeople {

    fun numRescueBoats(people: Array<Int>, limit: Int): Int {
        val sortedPeople = people.sorted()
        var smallest = 0
        var biggest = sortedPeople.size - 1

        var boats = 0

        while (biggest >= smallest) {
            if (sortedPeople[smallest] + sortedPeople[biggest] <= limit) {
                biggest -= 1
                smallest += 1
            } else {
                biggest -= 1
            }
            boats += 1
        }
        return boats
    }
}

fun main() {
    val boats = BoatsToSavePeople()
    val people = arrayOf(3, 5, 3, 4)
    val limit = 5
    println(boats.numRescueBoats(people, limit))
}